### PR TITLE
New Load Balancer object for HTTP Activity 4002

### PR DESCRIPTION
**Pull request by nathanbvail**
#### Description of changes: Updated from [Draft PR 861](https://github.com/ocsf/ocsf-schema/pull/861). 

- Created a new `Load Balancer` object that contains specific Load Balancer related information and added that object as an optional attribute in HTTP Activity 4002
- Created a new `endpoint_connections` object that sits as an attribute inside of the `load_balancer` object
- Created a new `endpoint` attribute that is an array of Network Endpoint objects
- Lastly, updated the dictionary with the necessary additions `endpoint`, `endpoint_connections`, and `load_balancer `

### CODE REVIEW

File: objects/load_balancer.json [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+ 
+    "metrics": {
+        "caption": "Metrics",
+        "description": "General purpose metrics associated with the load balancer.",
+        "is_array": true,
+        "requirement": "optional"
+        },
+    "dst_endpoint": { 
+        "caption": "Destination Endpoint",
+        "description": "The destination that the load balancer is distibuting traffic to.",
```
Comment (Aniak5) [inline objects/load_balancer.json:16]: Spelling error `distibuting` should be `distributing`.

Comment (Aniak5) [reply objects/load_balancer.json:16]: Also, we may want to reword as `The destination to which the load balancer is distributing traffic.`

File: objects/load_balancer.json [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    "message": {
+        "caption": "Message",
+        "description": "The load balancer message.",
+        "requirement": "optional"
+        }, 
+    "name": {
+        "description": "The name of the load balancer."
+        },
+    "uid": {
+        "description": "The unique identfier for the load balancer."
```
Comment (Aniak5) [inline objects/load_balancer.json:53]: Misspelling in `identfier`. should be `identifier`

File: objects/endpoint_connections.json [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    "attributes": {
+        "code": { 
+            "caption": "Response Code",
+            "description": "A numeric response status code detailing the connection.",
+            "requirement": "optional"
+            },
+        "endpoint": {  
+            "caption": "Endpoint Array", 
+            "description": "An array of endpoint/s that fielded the connection attempt.",
+            "requirement": "optional"
```
Comment (Aniak5) [inline objects/endpoint_connections.json:15]: We should probably make the `requirement` on both of these attributes to be `recommended` instead of `optional`.

File: objects/endpoint_connections.json [hunk 4]
```diff
@@ -0,0 +1,26 @@
+{
+    "caption": "Endpoint Connections",
+    "name": "endpoint_connections",
+    "description": "The Endpoint Connections object contains information regarding the connections to multiple destination endpoints.",
+    "extends": "object",
+    "attributes": {
+        "code": { 
+            "caption": "Response Code",
+            "description": "A numeric response status code detailing the connection.",
```
Comment (Aniak5) [inline objects/endpoint_connections.json:9]: We may want to reword this to be `A numerical response status code providing details about the connection.`.

File: dictionary.json [hunk 5]
```diff
@@ -1343,6 +1343,18 @@
...
+      "caption": "Endpoints",
+      "description": "The Endpoint/s related to the entity.",
+      "is_array": true,
+      "type": "network_endpoint"
+    },
+    "endpoint_connections": {
+      "caption": "Endpoint Connections",
+      "description": "Connection attempts and their responses. See specific usage.",
+      "is_array": true,
+      "type": "endpoint_connections"
```
Comment (Aniak5) [inline dictionary.json:1356]: This should become singular per the comments on that specific object.

File: dictionary.json [hunk 6]
```diff
@@ -1343,6 +1343,18 @@
       "description": "The end time of a time period. See specific usage.",
       "type": "timestamp_t"
     },
+    "endpoint": {
```
Comment (Aniak5) [inline dictionary.json:1346]: This is no longer needed and can be deleted.

File: objects/load_balancer.json [hunk 7]
```diff
@@ -0,0 +1,57 @@
+{
+    "caption": "Load Balancer",
+    "name": "load_balancer",
```
Comment (Aniak5) [inline objects/load_balancer.json:3]: Something should be required within this object so that we don't wind up with an empty `load_balancer` object. At the very least  `name` and `uid` should  be added to a `at_least_one` constraint. Though we might want one or both of those to be required.

Comment (nathanbvail) [reply objects/load_balancer.json:3]: By extending `_entity`, it automatically adds `name` or `uid` as a constraint, I included those attributes in the object so I could update the descriptions to be LB specific 
image

Comment (Aniak5) [reply objects/load_balancer.json:3]: Missed that, great catch!

File: objects/endpoint_connections.json [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+            "requirement": "optional"
+            },
+        "endpoint": {  
+            "caption": "Endpoint Array", 
+            "description": "An array of endpoint/s that fielded the connection attempt.",
```
[the remark this answers is not in the document]
Comment (nathanbvail) [reply objects/endpoint_connections.json:14]: Per todays meetings (personal notes): Remove `endpoint` from the dictionary & replace `endpoint` attribute with the existing `network_endpoint` attribute inside of `endpoint_connections.json` .

**Comment by nathanbvail** [comment]
Comments from 12/13 meeting: 
- Create a LB profile that contains the LB object, this will allow for additional Load Balancer types to utilize the objects across classes (i.e. HTTP Activity & Network Activity classes) 
- Remove `endpoint` from the dictionary & replace `endpoint` attribute with the existing `network_endpoint` attribute inside of `endpoint_connections.json`

**Comment by nathanbvail** [comment]
@Aniak5 / @pagbabian-splunk - Before I push changes , are we all in agreement on going forward with a `Load Balancer` Profile that can be applied to the Networking class (expanding to other classes if needed) ? 

If that is the case then I will remove the `load_balancer` attribute from `http.json` and transform this PR to just be the object.

**Comment by Aniak5** [comment]
> @Aniak5 / @pagbabian-splunk - Before I commit changes , are we all in agreement on going forward with a `Load Balancer` Profile that can be applied to the Networking class (expanding to other classes if needed) ?
> 
> If that is the case then I will remove the `load_balancer` attribute from `http.json` and transform this PR to just be the object.

Yup, that's exactly what we want to do @nathanbvail.

**Comment by nathanbvail** [comment]
Uploading screenshots of the LB Profile changes - image 
image

**Review by mikeradka — APPROVED** [review]

**Comment by nathanbvail** [comment]
Per todays OCSF General Call (12/19) - I will be closing out this PR & creating a new PR due to the profile approach instead of an object.
